### ISSUE TITLE
Target returned by `yarray_event_target` cannot be used for querying data

### ISSUE THREAD

**Issue by ryanbinnsrv** [y-crdt/y-crdt#384]
I'm attempting to use the C++ FFI interface and all seems to be working ok except for the observe callbacks. I'm adding an element to an array and querying data from the `yarray_observe` callback. My callback code is this:
```c++
void callback(void *state, const YArrayEvent *event)
{
    Branch *target = yarray_event_target(event);
    uint32_t length = yarray_len(target);
    std::cout << length << std::endl;
}
```
The callback is being called properly, but the value being printed out is nonsense (I'm getting random numbers like 2358227345 when the length should be 4).

I added this code in `yarray_event_target` and it prints the correct length, so as far as this it is correct:
```rust
pub unsafe extern "C" fn yarray_event_target(e: *const YArrayEvent) -> *mut Branch {
    assert!(!e.is_null());
    let out = (&*e).target().clone();
    println!("Array length (yarray_event_target): {}", out.len((&*e).txn())); // <----
    Box::into_raw(Box::new(out)) as *mut _
}
```

I think the reason is the return value of `yarray_event_target`. Comparing with `ytext_event_target` immediately above, it is returning this:
```rust
out.into_raw_branch()
```
I tried changing `yarray_event_target` to return `out.into_raw_branch()`, and my callback is now working correctly. Is this the correct fix? I notice that `ymap_event_target`, `yxmlelem_event_target` and `yxmltext_event_target` are currently the same as `yarray_event_target` (`ytext_event_target` is the only one that is different).

**Comment by Horusiath** [comment]
Hi @ryanbinnsrv do you have some code to reproduce the issue? It would help us to fix this issue faster. Currently one of the things we're focusing on is related to fixing invalid pointer issues.

**Comment by ryanbinnsrv** [comment]
Yes, this reproduces it every run.
```c++
#include <cassert>
#include <libyrs.h>

int main(int argc, char *argv[])
{
    YDoc *doc = ydoc_new();

    const Branch *array = yarray(doc, "array1");

    uint32_t subscription = yarray_observe(
        array,
        nullptr,
        [](void *state, const YArrayEvent *event)
        {
            const Branch *target = yarray_event_target(event);

            assert(yarray_len(target) == 1);
        });

    YTransaction *txn = ydoc_write_transaction(doc, 0, nullptr);

    YInput item{ Y_JSON_NUM, 1, { .num = 25.0 } };

    yarray_insert_range(array, txn, 0, &item, 1);

    ytransaction_commit(txn);

    yarray_unobserve(array, subscription);

    ydoc_destroy(doc);

    return 0;
}
```

**Comment by Horusiath** [comment]
Technically this issue should already be fixed in v0.18+. I've added the tests to main repo confirm that.
